csrc/tensor_metadata.cpp

```diff
@@ -323,30 +323,63 @@ inferAndValidateAllocationSizesAndStrides(
   return {std::move(sizes), std::move(strides)};
 }
 
+namespace {
+std::pair<std::vector<int64_t>, std::vector<int64_t>> unshardedSizesAndStrides(
+    TensorView* tv,
+    c10::IntArrayRef sizes,
+    c10::IntArrayRef strides) {
+  std::vector<int64_t> unsharded_sizes(sizes.size());
+  std::vector<int64_t> unsharded_strides(strides.size());
+  for (const auto i : c10::irange(sizes.size())) {
+    if (tv->getLogicalDomain()[i]->isDeviceDim()) {
+      unsharded_sizes[i] = tv->getDeviceMesh().size();
```

Comment:
I think this is another one that would break with multidimensional device meshes.